README.md
# NAME

Starwoman - because Starman does the same thing over and over again expecting different results

# VERSION

version 0.001

# SYNOPSIS

    # Run app.psgi with the default settings
    > starwoman

    # run with Server::Starter
    > start_server --port 127.0.0.1:80 -- starwoman --workers 32 myapp.psgi

    # UNIX domain sockets
    > starwoman --listen /tmp/starwoman.sock

Read more options and configurations by running \`perldoc starwoman\` (lower-case s).

# DESCRIPTION

Starwoman is a PSGI perl web server that has unique features such as:

- High Performance

    Uses the fast XS/C HTTP header parser

- Preforking

    Spawns workers preforked like most high performance UNIX servers
    do. Starwoman also reaps dead children and automatically restarts the
    worker pool.

- Signals

    Supports `HUP` for graceful worker restarts, and `TTIN`/`TTOU` to
    dynamically increase or decrease the number of worker processes, as
    well as `QUIT` to gracefully shutdown the worker processes.

- Superdaemon aware

    Supports [Server::Starter](https://metacpan.org/pod/Server::Starter) for hot deploy and graceful restarts.

- Multiple interfaces and UNIX Domain Socket support

    Able to listen on multiple interfaces including UNIX sockets.

- Small memory footprint

    Preloading the applications with `--preload-app` command line option
    enables copy-on-write friendly memory management. Also, the minimum
    memory usage Starwoman requires for the master process is 7MB and
    children (workers) is less than 3.0MB.

- PSGI compatible

    Can run any PSGI applications and frameworks

- HTTP/1.1 support

    Supports chunked requests and responses, keep-alive and pipeline requests.

- UNIX only

    This server does not support Win32.

# PERFORMANCE

Here's a simple benchmark using `Hello.psgi`.

    -- server: Starwoman (workers=10)
    Requests per second:    6849.16 [#/sec] (mean)
    -- server: Twiggy
    Requests per second:    3911.78 [#/sec] (mean)
    -- server: AnyEvent::HTTPD
    Requests per second:    2738.49 [#/sec] (mean)
    -- server: HTTP::Server::PSGI
    Requests per second:    2218.16 [#/sec] (mean)
    -- server: HTTP::Server::PSGI (workers=10)
    Requests per second:    2792.99 [#/sec] (mean)
    -- server: HTTP::Server::Simple
    Requests per second:    1435.50 [#/sec] (mean)
    -- server: Corona
    Requests per second:    2332.00 [#/sec] (mean)
    -- server: POE
    Requests per second:    503.59 [#/sec] (mean)

This benchmark was processed with `ab -c 10 -t 1 -k` on MacBook Pro
13" late 2009 model on Mac OS X 10.6.2 with perl 5.10.0. YMMV.

# NOTES

Because Starwoman runs as a preforking model, it is not recommended to
serve the requests directly from the internet, especially when slow
requesting clients are taken into consideration. It is suggested to
put Starwoman workers behind the frontend servers such as nginx, and use
HTTP proxy with TCP or UNIX sockets.

This is a fork of Starman for one particular reason: to stop the endless
forking of immediately dying children when app.psgi can't be loaded,
which flooded the log file and pegged the CPU. Starman hasn't been
maintained, hence the fork. Do not assume I will be any better about
maintaining this, considering how much attention I give to projects I
wrote myself.

# AUTHOR

Ashley Willis <awillis@synacor.com>

Tatsuhiko Miyagawa <miyagawa@bulknews.net> wrote [Starman](https://metacpan.org/pod/Starman), which this module
is a fork of with minimal modifications.

Andy Grundman wrote [Catalyst::Engine::HTTP::Prefork](https://metacpan.org/pod/Catalyst::Engine::HTTP::Prefork), which this module
is heavily based on.

Kazuho Oku wrote [Net::Server::SS::PreFork](https://metacpan.org/pod/Net::Server::SS::PreFork) that makes it easy to add
[Server::Starter](https://metacpan.org/pod/Server::Starter) support to this software.

# COPYRIGHT

Ashley Willis, 2019
Tatsuhiko Miyagawa, 2010-

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

[Plack](https://metacpan.org/pod/Plack) [Catalyst::Engine::HTTP::Prefork](https://metacpan.org/pod/Catalyst::Engine::HTTP::Prefork) [Net::Server::PreFork](https://metacpan.org/pod/Net::Server::PreFork)


Changes
Revision history for Perl extension Starwoman

{{$NEXT}}

0.001  2019-04-29 13:20:06 EDT
        - fork Starman to Starwoman
        - stop endless forking of immediately dying children, which floods the log file and pegs the cpu (https://github.com/miyagawa/Starman/pull/117)

0.4014  2015-06-03 12:01:00 PDT
        - Treat ECONNRESET like EPIPE (i.e. ignore), not as a fatal error #114 (Tim Bunce)

0.4013  2015-05-14 15:01:20 PDT
        - Fixed some bad git merges.

0.4012  2015-05-14 14:59:48 PDT
        - Add --net_server-* options to pass directly to Net::Server backend (#109)
        - Updated documentation

0.4011  2014-11-11 08:07:43 PST
        - Move the app dispatch into a method #107

0.4010  2014-08-22 09:37:22 PDT
        - Support --read-timeout #103 (slobo)
        - Handle Expect header case insensitively #101 (oschwald)

0.4009  2014-04-03 14:39:27 PDT
        - Do not send chunked body for HEAD requests #87 (therigu)
        - Added --disable-proctitle option to disable the proctitle change #97

0.4008  2013-09-08 21:09:22 PDT
        - Make response write loop a zero-copy (ap)

0.4007  2013-09-02 17:11:38 PDT
        - Handle EPIPE and stops writing to the socket #84 (ap)

0.4006  2013-08-16 12:43:19 PDT
        - Same as 0.4005, non-devel

0.4005  2013-08-13 22:12:11 PDT
        - Fix SSL implementation bug where body longer than 16K doesn't get written correctly #78 (kazeburo, siracusa) 

0.4004  2013-08-12 11:41:13 PDT
        - Note that SSL support is experimental

0.4003  2013-08-08 14:32:24 PDT
        - Fix ssl condition for banner

0.4002  2013-08-08 14:20:45 PDT
        - Fix banner message for host/port broken in 0.4 (siracusa)

0.4001  2013-07-29 23:14:10 PDT
        - Skip ssl tests if LWP doesn't support HTTPS

0.4000  2013-07-28 23:53:55 PDT
        - support SSL (aristotle)

0.3014  2013-06-16 01:07:52 PDT
        - bump Test::TCP dep, not really necessary but avoid 1.27 bug

0.3013  2013-06-12 22:52:54 PDT
        - Fix writer tests for Plack update

0.3012  2013-06-12 12:31:11 PDT
        - typo fixes
        - Fix dependencies for LWP::UserAgent

0.3011  2013-04-24 17:39:31 PDT
        - Disabled flock serialization when it's unnecessary. This will improve the performance
          when you have many Starman worker processes on Linux systems (kazeburo) #69

0.3010  2013-04-24 07:04:02 PDT
        - Switch back to Module::Build::Tiny with fixed #!perl shebang

0.3009  2013-04-23 17:34:57 PDT
        - Switch to MakeMaker because of shebang bug
          https://github.com/Leont/module-build-tiny/issues/3

0.3008  2013-04-06 22:04:53 PDT
        - switch to Module::Build::Tiny with Milla. Might not install bat correctly but we don't
          support Win32 anyway.

0.3007  2013-03-28 12:55:45 PDT
        - Accept HTTP requests with LF termination per HTTP spec recommendation (oschwald) #56
        - Documentation fix (oalders)

0.3006  Wed Dec 19 09:55:05 JST 2012
        - Clear out @ARGV, rather than restoring it, to avoid messing with Net::Server internals

0.3005  Wed Nov 14 19:46:31 PST 2012
        - Added a warning in runtime/documentation to NOT use -r/-R with Starman

0.3004  Thu Nov  8 19:40:45 PST 2012
        - Added --interval option to the sample start_server command
        - Makefile.PL fix

0.3003  Thu Sep 27 09:39:56 JST 2012
        - Fixed the test hang in some environments, introduced in 0.3002 [RT:79865]

0.3002  Tue Sep 25 15:26:43 JST 2012
        - Added a documentation for --signal-on-term for Server::Starter 0.12 (kazuho, ether)
        - Set REMOTE_PORT PSGI environment variable #50 (dex4er)
        - Fix a test failure with a directory containing whitespace (clkao)

0.3001  Mon Jun 25 10:57:20 PDT 2012
        - Fix SERVER_NAME and SERVER_PORT not exist on UNIX socket mode #24
        - Improved documentation
        - Ensure that chunk buffer contains terminating HTTP newline (Peter Makholm)

0.3000  Mon Feb 20 16:31:44 PST 2012
        - This be a 0.3 release

0.29_90 Thu Dec  1 19:40:52 PST 2011
        - Changed the way server handles HUP and QUIT signals
          HUP will just restart all the workers gracefully
          QUIT will gracefully shutdown workers and the master
          See `man 1 starman` and look for SIGNALS section.

0.2014  Sun Sep 18 12:43:06 PDT 2011
        - Fixed broken PSGI response headers after the output (cho45)

0.2013  Sat Jun 25 11:51:47 PDT 2011
        - Relaxed the harakiri tests (audreyt)

0.2012  Wed Jun 22 13:51:59 PDT 2011
        - Implemented psgix.harakiri mode (audreyt)
        - Added --error-log option (Paulo E. Castro)

0.2011  Tue May 24 09:41:52 PDT 2011
        - Fix chunked response with 0-length PSGI array elements (chmrr)

0.2010  Mon Mar 28 16:23:23 PDT 2011
        - Fixed packaging. No changes.

0.2009  Fri Mar 25 19:15:23 PDT 2011
        - Requires Plack 0.9971 to support localizing $0 to fix the FindBin issues #7, #15, #18, #19
        - Calls srand() automatically in the child init hook to avoid a fixed random seed #20
        - Implemented --keepalive-timeout which defaults to 1 (acme)

0.2008  Mon Feb 14 17:19:20 PST 2011
        - Documented that -E is automatically set to 'deployment' RT:61517 (timbunce)
        - Check the defined-ness of the input buffer to suppress warnings RT:60007

0.2007 Thu Sep 30 14:09:00 PDT 2010
        - Fixed a bug where Date header can be duplicate if the app generates one (spleenjack)

0.2006 Fri Jul  2 17:21:22 PDT 2010
        - Fixed a bug in chunked response when Content-Length is 0. #8 (chiba)
        - Documented --pid and --daemonize

0.2005 Fri Jul  2 17:02:16 PDT 2010
        - Don't use lib 'lib'
        - Documentation updates (miyagawa, grantm)

0.2004 Tue Apr 20 21:22:31 JST 2010
        - Delay set $0 in master so FindBin works. #7

0.2003 Mon Apr 19 15:19:06 JST 2010
        - Upped Plack dependency


lib/Starwoman.pm
package Starwoman;

use strict;
use 5.008_001;

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Starwoman - because Starman does the same thing over and over again expecting different results

=head1 SYNOPSIS

  # Run app.psgi with the default settings
  > starwoman

  # run with Server::Starter
  > start_server --port 127.0.0.1:80 -- starwoman --workers 32 myapp.psgi

  # UNIX domain sockets
  > starwoman --listen /tmp/starwoman.sock

Read more options and configurations by running `perldoc starwoman` (lower-case s).

=head1 DESCRIPTION

Starwoman is a PSGI perl web server that has unique features such as:

=over 4

=item High Performance

Uses the fast XS/C HTTP header parser

=item Preforking

Spawns workers preforked like most high performance UNIX servers
do. Starwoman also reaps dead children and automatically restarts the
worker pool.

=item Signals

Supports C<HUP> for graceful worker restarts, and C<TTIN>/C<TTOU> to
dynamically increase or decrease the number of worker processes, as
well as C<QUIT> to gracefully shutdown the worker processes.

=item Superdaemon aware

Supports L<Server::Starter> for hot deploy and graceful restarts.

=item Multiple interfaces and UNIX Domain Socket support

Able to listen on multiple interfaces including UNIX sockets.

=item Small memory footprint

Preloading the applications with C<--preload-app> command line option
enables copy-on-write friendly memory management. Also, the minimum
memory usage Starwoman requires for the master process is 7MB and
children (workers) is less than 3.0MB.

=item PSGI compatible

Can run any PSGI applications and frameworks

=item HTTP/1.1 support

Supports chunked requests and responses, keep-alive and pipeline requests.

=item UNIX only

This server does not support Win32.

=back

=head1 PERFORMANCE

Here's a simple benchmark using C<Hello.psgi>.

  -- server: Starwoman (workers=10)
  Requests per second:    6849.16 [#/sec] (mean)
  -- server: Twiggy
  Requests per second:    3911.78 [#/sec] (mean)
  -- server: AnyEvent::HTTPD
  Requests per second:    2738.49 [#/sec] (mean)
  -- server: HTTP::Server::PSGI
  Requests per second:    2218.16 [#/sec] (mean)
  -- server: HTTP::Server::PSGI (workers=10)
  Requests per second:    2792.99 [#/sec] (mean)
  -- server: HTTP::Server::Simple
  Requests per second:    1435.50 [#/sec] (mean)
  -- server: Corona
  Requests per second:    2332.00 [#/sec] (mean)
  -- server: POE
  Requests per second:    503.59 [#/sec] (mean)

This benchmark was processed with C<ab -c 10 -t 1 -k> on MacBook Pro
13" late 2009 model on Mac OS X 10.6.2 with perl 5.10.0. YMMV.

=head1 NOTES

Because Starwoman runs as a preforking model, it is not recommended to
serve the requests directly from the internet, especially when slow
requesting clients are taken into consideration. It is suggested to
put Starwoman workers behind the frontend servers such as nginx, and use
HTTP proxy with TCP or UNIX sockets.

This is a fork of Starman for one particular reason: to stop the endless
forking of immediately dying children when app.psgi can't be loaded,
which flooded the log file and pegged the CPU. Starman hasn't been
maintained, hence the fork. Do not assume I will be any better about
maintaining this, considering how much attention I give to projects I
wrote myself.

=head1 AUTHOR

Ashley Willis E<lt>awillis@synacor.comE<gt>

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt> wrote L<Starman>, which this module
is a fork of with minimal modifications.

Andy Grundman wrote L<Catalyst::Engine::HTTP::Prefork>, which this module
is heavily based on.

Kazuho Oku wrote L<Net::Server::SS::PreFork> that makes it easy to add
L<Server::Starter> support to this software.


=head1 COPYRIGHT

Ashley Willis, 2019
Tatsuhiko Miyagawa, 2010-

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<Plack> L<Catalyst::Engine::HTTP::Prefork> L<Net::Server::PreFork>

=cut


lib/Starwoman/Server.pm
package Starwoman::Server;

use strict;
use base 'Net::Server::PreFork';

use Data::Dump qw(dump);
use Socket qw(IPPROTO_TCP TCP_NODELAY);
use IO::Socket qw(:crlf);
use HTTP::Parser::XS qw(parse_http_request);
use HTTP::Status qw(status_message);
use HTTP::Date qw(time2str);
use POSIX qw(EINTR EPIPE ECONNRESET);
use Symbol;
use Try::Tiny;

use Plack::Util;
use Plack::TempBuffer;

use constant DEBUG        => $ENV{STARWOMAN_DEBUG} || 0;
use constant CHUNKSIZE    => 64 * 1024;

my $null_io = do { open my $io, "<", \""; $io };

use Net::Server::SIG qw(register_sig);

# Override Net::Server's HUP handling - just restart all the workers and that's about it
sub sig_hup {
    my $self = shift;
    $self->hup_children;
}

sub run {
    my($self, $app, $options) = @_;

    $self->{app} = $app;
    $self->{options} = $options;

    my %extra = ();

    if ($options->{net_server_args}) {
        %extra = %{ $options->{net_server_args} };
    }

    if ( $options->{pid} ) {
        $extra{pid_file} = $options->{pid};
    }
    if ( $options->{daemonize} ) {
        $extra{setsid} = $extra{background} = 1;
    }
    if ( $options->{error_log} ) {
        $extra{log_file} = $options->{error_log};
    }
    if ( DEBUG ) {
        $extra{log_level} = 4;
    }
    if ( $options->{ssl_cert} ) {
        $extra{SSL_cert_file} = $options->{ssl_cert};
    }
    if ( $options->{ssl_key} ) {
        $extra{SSL_key_file} = $options->{ssl_key};
    }
    if (! exists $options->{keepalive}) {
        $options->{keepalive} = 1;
    }
    if (! exists $options->{keepalive_timeout}) {
        $options->{keepalive_timeout} = 1;
    }
    if (! exists $options->{read_timeout}) {
        $options->{read_timeout} = 5;
    }
    if (! exists $options->{proctitle}) {
        $options->{proctitle} = 1;
    }

    my @port;
    for my $listen (@{$options->{listen} || [ "$options->{host}:$options->{port}" ]}) {
        my %listen;
        if ($listen =~ /:/) {
            my($h, $p, $opt) = split /:/, $listen, 3;
            $listen{host} = $h if $h;
            $listen{port} = $p;
            $listen{proto} = 'ssl' if 'ssl' eq lc $opt;
        } else {
            %listen = (
                host  => 'localhost',
                port  => $listen,
                proto => 'unix',
            );
        }
        push @port, \%listen;
    }

    my $workers = $options->{workers} || 5;
    local @ARGV = ();

    $self->SUPER::run(
        port                => \@port,
        host                => '*',   # default host
        proto               => $options->{ssl} ? 'ssl' : 'tcp', # default proto
        serialize           => ( $^O =~ m!(linux|darwin|bsd|cygwin)$! ) ? 'none' : 'flock',
        min_servers         => $options->{min_servers}       || $workers,
        min_spare_servers   => $options->{min_spare_servers} || $workers - 1,
        max_spare_servers   => $options->{max_spare_servers} || $workers - 1,
        max_servers         => $options->{max_servers}       || $workers,
        max_requests        => $options->{max_requests}      || 1000,
        user                => $options->{user}              || $>,
        group               => $options->{group}             || $),
        listen              => $options->{backlog}           || 1024,
        check_for_waiting   => 1,
        no_client_stdout    => 1,
        %extra
    );
}

sub pre_loop_hook {
    my $self = shift;

    my $port = $self->{server}->{port}->[0];
    my $proto = $port->{proto} eq 'ssl'  ? 'https' :
                $port->{proto} eq 'unix' ? 'unix'  :
                                           'http';

    $self->{options}{server_ready}->({
        host => $port->{host},
        port => $port->{port},
        proto => $proto,
        server_software => 'Starwoman',
    }) if $self->{options}{server_ready};

    register_sig(
        TTIN => sub { $self->{server}->{$_}++ for qw( min_servers max_servers ) },
        TTOU => sub { $self->{server}->{$_}-- for qw( min_servers max_servers ) },
        QUIT => sub { $self->server_close(1) },
    );
}

sub server_close {
    my($self, $quit) = @_;

    if ($quit) {
        $self->log(2, $self->log_time . " Received QUIT. Running a graceful shutdown\n");
        $self->{server}->{$_} = 0 for qw( min_servers max_servers );
        $self->hup_children;
        while (1) {
            Net::Server::SIG::check_sigs();
            $self->coordinate_children;
            last if !keys %{$self->{server}{children}};
            sleep 1;
        }
        $self->log(2, $self->log_time . " Worker processes cleaned up\n");
    }

    $self->SUPER::server_close();
}

sub run_parent {
    my $self = shift;
    $0 = "starwoman master " . join(" ", @{$self->{options}{argv} || []})
        if $self->{options}{proctitle};
    no warnings 'redefine';
    local *Net::Server::PreFork::register_sig = sub {
        my %args = @_;
        delete $args{QUIT};
        Net::Server::SIG::register_sig(%args);
    };
    $self->SUPER::run_parent(@_);
}

# The below methods run in the child process

sub child_init_hook {
    my $self = shift;
    srand();
    if ($self->{options}->{psgi_app_builder}) {
        DEBUG && warn "[$$] Initializing the PSGI app\n";
        # try/catch based on https://github.com/miyagawa/Starman/pull/117
        try {
            $self->{app} = $self->{options}->{psgi_app_builder}->();
        } catch {
            die $_ unless $_ =~ /Error while loading/;
            DEBUG && warn "[$$] Failed to load the app, will try again on request: $_\n";
        };
    }
    $0 = "starwoman worker " . join(" ", @{$self->{options}{argv} || []})
        if $self->{options}{proctitle};

}

sub post_accept_hook {
    my $self = shift;

    $self->{client} = {
        headerbuf => '',
        inputbuf  => '',
        keepalive => 1,
    };
}

sub dispatch_request {
    my ($self, $env) = @_;

    # Run PSGI apps
    my $res = Plack::Util::run_app($self->{app}, $env);

    if (ref $res eq 'CODE') {
        $res->(sub { $self->_finalize_response($env, $_[0]) });
    } else {
        $self->_finalize_response($env, $res);
    }
}

sub process_request {
    my $self = shift;
    my $conn = $self->{server}->{client};

    if ($conn->NS_proto eq 'TCP') {
        setsockopt($conn, IPPROTO_TCP, TCP_NODELAY, 1)
            or die $!;
    }

    while ( $self->{client}->{keepalive} ) {
        last if !$conn->connected;

        # Read until we see all headers
        last if !$self->_read_headers;

        my $env = {
            REMOTE_ADDR     => $self->{server}->{peeraddr},
            REMOTE_HOST     => $self->{server}->{peerhost} || $self->{server}->{peeraddr},
            REMOTE_PORT     => $self->{server}->{peerport} || 0,
            SERVER_NAME     => $self->{server}->{sockaddr} || 0, # XXX: needs to be resolved?
            SERVER_PORT     => $self->{server}->{sockport} || 0,
            SCRIPT_NAME     => '',
            'psgi.version'      => [ 1, 1 ],
            'psgi.errors'       => *STDERR,
            'psgi.url_scheme'   => ($conn->NS_proto eq 'SSL' ? 'https' : 'http'),
            'psgi.nonblocking'  => Plack::Util::FALSE,
            'psgi.streaming'    => Plack::Util::TRUE,
            'psgi.run_once'     => Plack::Util::FALSE,
            'psgi.multithread'  => Plack::Util::FALSE,
            'psgi.multiprocess' => Plack::Util::TRUE,
            'psgix.io'          => $conn,
            'psgix.input.buffered' => Plack::Util::TRUE,
            'psgix.harakiri' => Plack::Util::TRUE,
        };

        # Parse headers
        my $reqlen = parse_http_request(delete $self->{client}->{headerbuf}, $env);
        if ( $reqlen == -1 ) {
            # Bad request
            DEBUG && warn "[$$] Bad request\n";
            $self->_http_error(400, { SERVER_PROTOCOL => "HTTP/1.0" });
            last;
        }

        # Initialize PSGI environment
        # Determine whether we will keep the connection open after the request
        my $connection = delete $env->{HTTP_CONNECTION};
        my $proto = $env->{SERVER_PROTOCOL};
        if ( $proto && $proto eq 'HTTP/1.0' ) {
            if ( $connection && $connection =~ /^keep-alive$/i ) {
                # Keep-alive only with explicit header in HTTP/1.0
                $self->{client}->{keepalive} = 1;
            }
            else {
                $self->{client}->{keepalive} = 0;
            }
        }
        elsif ( $proto && $proto eq 'HTTP/1.1' ) {
            if ( $connection && $connection =~ /^close$/i ) {
                $self->{client}->{keepalive} = 0;
            }
            else {
                # Keep-alive assumed in HTTP/1.1
                $self->{client}->{keepalive} = 1;
            }

            # Do we need to send 100 Continue?
            if ( $env->{HTTP_EXPECT} ) {
                if ( lc $env->{HTTP_EXPECT} eq '100-continue' ) {
                    _syswrite($conn, \('HTTP/1.1 100 Continue' . $CRLF . $CRLF));
                    DEBUG && warn "[$$] Sent 100 Continue response\n";
                }
                else {
                    DEBUG && warn "[$$] Invalid Expect header, returning 417\n";
                    $self->_http_error( 417, $env );
                    last;
                }
            }

            unless ($env->{HTTP_HOST}) {
                # No host, bad request
                DEBUG && warn "[$$] Bad request, HTTP/1.1 without Host header\n";
                $self->_http_error( 400, $env );
                last;
            }
        }

        unless ($self->{options}->{keepalive}) {
            DEBUG && warn "[$$] keep-alive is disabled. Closing the connection after this request\n";
            $self->{client}->{keepalive} = 0;
        }

        $self->_prepare_env($env);

        $self->dispatch_request($env);

        DEBUG && warn "[$$] Request done\n";

        if ( $self->{client}->{keepalive} ) {
            # If we still have data in the input buffer it may be a pipelined request
            if ( $self->{client}->{inputbuf} ) {
                if ( $self->{client}->{inputbuf} =~ /^(?:GET|HEAD)/ ) {
                    if ( DEBUG ) {
                        warn "Pipelined GET/HEAD request in input buffer: "
                            . dump( $self->{client}->{inputbuf} ) . "\n";
                    }

                    # Continue processing the input buffer
                    next;
                }
                else {
                    # Input buffer just has junk, clear it
                    if ( DEBUG ) {
                        warn "Clearing junk from input buffer: "
                            . dump( $self->{client}->{inputbuf} ) . "\n";
                    }

                    $self->{client}->{inputbuf} = '';
                }
            }

            DEBUG && warn "[$$] Waiting on previous connection for keep-alive request...\n";

            my $sel = IO::Select->new($conn);
            last unless $sel->can_read($self->{options}->{keepalive_timeout});
        }
    }

    DEBUG && warn "[$$] Closing connection\n";
}

sub _read_headers {
    my $self = shift;

    eval {
        local $SIG{ALRM} = sub { die "Timed out\n"; };

        alarm( $self->{options}->{read_timeout} );

        while (1) {
            # Do we have a full header in the buffer?
            # This is before sysread so we don't read if we have a pipelined request
            # waiting in the buffer
            last if defined $self->{client}->{inputbuf} && $self->{client}->{inputbuf} =~ /$CR?$LF$CR?$LF/s;

            # If not, read some data
            my $read = sysread $self->{server}->{client}, my $buf, CHUNKSIZE;

            if ( !defined $read || $read == 0 ) {
                die "Read error: $!\n";
            }

            if ( DEBUG ) {
                warn "[$$] Read $read bytes: " . dump($buf) . "\n";
            }

            $self->{client}->{inputbuf} .= $buf;
        }
    };

    alarm(0);

    if ( $@ ) {
        if ( $@ =~ /Timed out/ ) {
            DEBUG && warn "[$$] Client connection timed out\n";
            return;
        }

        if ( $@ =~ /Read error/ ) {
            DEBUG && warn "[$$] Read error: $!\n";
            return;
        }
    }

    # Pull out the complete header into a new buffer
    $self->{client}->{headerbuf} = $self->{client}->{inputbuf};

    # Save any left-over data, possibly body data or pipelined requests
    $self->{client}->{inputbuf} =~ s/.*?$CR?$LF$CR?$LF//s;

    return 1;
}

sub _http_error {
    my ( $self, $code, $env ) = @_;

    my $status = $code || 500;
    my $msg    = status_message($status);

    my $res = [
        $status,
        [ 'Content-Type' => 'text/plain', 'Content-Length' => length($msg) ],
        [ $msg ],
    ];

    $self->{client}->{keepalive} = 0;
    $self->_finalize_response($env, $res);
}

sub _prepare_env {
    my($self, $env) = @_;

    my $get_chunk = sub {
        if ($self->{client}->{inputbuf}) {
            my $chunk = delete $self->{client}->{inputbuf};
            return ($chunk, length $chunk);
        }
        my $read = sysread $self->{server}->{client}, my($chunk), CHUNKSIZE;
        return ($chunk, $read);
    };

    my $chunked = do { no warnings; lc delete $env->{HTTP_TRANSFER_ENCODING} eq 'chunked' };

    if (my $cl = $env->{CONTENT_LENGTH}) {
        my $buf = Plack::TempBuffer->new($cl);
        while ($cl > 0) {
            my($chunk, $read) = $get_chunk->();

            if ( !defined $read || $read == 0 ) {
                die "Read error: $!\n";
            }

            $cl -= $read;
            $buf->print($chunk);
        }
        $env->{'psgi.input'} = $buf->rewind;
    } elsif ($chunked) {
        my $buf = Plack::TempBuffer->new;
        my $chunk_buffer = '';
        my $length;

    DECHUNK:
        while (1) {
            my($chunk, $read) = $get_chunk->();
            $chunk_buffer .= $chunk;

            while ( $chunk_buffer =~ s/^(([0-9a-fA-F]+).*\015\012)// ) {
                my $trailer   = $1;
                my $chunk_len = hex $2;

                if ($chunk_len == 0) {
                    last DECHUNK;
                } elsif (length $chunk_buffer < $chunk_len + 2) {
                    $chunk_buffer = $trailer . $chunk_buffer;
                    last;
                }

                $buf->print(substr $chunk_buffer, 0, $chunk_len, '');
                $chunk_buffer =~ s/^\015\012//;

                $length += $chunk_len;
            }

            last unless $read && $read > 0;
        }

        $env->{CONTENT_LENGTH} = $length;
        $env->{'psgi.input'}   = $buf->rewind;
    } else {
        $env->{'psgi.input'} = $null_io;
    }
}

sub _finalize_response {
    my($self, $env, $res) = @_;

    if ($env->{'psgix.harakiri.commit'}) {
        $self->{client}->{keepalive} = 0;
        $self->{client}->{harakiri} = 1;
    }

    my $protocol = $env->{SERVER_PROTOCOL};
    my $status   = $res->[0];
    my $message  = status_message($status);

    my(@headers, %headers);
    push @headers, "$protocol $status $message";

    # Switch on Transfer-Encoding: chunked if we don't know Content-Length.
    my $chunked;
    my $headers = $res->[1];
    for (my $i = 0; $i < @$headers; $i += 2) {
        my $k = $headers->[$i];
        my $v = $headers->[$i + 1];
        next if $k eq 'Connection';
        push @headers, "$k: $v";
        $headers{lc $k} = $v;
    }

    if ( $protocol eq 'HTTP/1.1' ) {
        if ( !exists $headers{'content-length'} ) {
            if ( $status !~ /^1\d\d|[23]04$/ && $env->{REQUEST_METHOD} ne 'HEAD' ) {
                DEBUG && warn "[$$] Using chunked transfer-encoding to send unknown length body\n";
                push @headers, 'Transfer-Encoding: chunked';
                $chunked = 1;
            }
        }
        elsif ( my $te = $headers{'transfer-encoding'} ) {
            if ( $te eq 'chunked' ) {
                DEBUG && warn "[$$] Chunked transfer-encoding set for response\n";
                $chunked = 1;
            }
        }
    } else {
        if ( !exists $headers{'content-length'} ) {
            DEBUG && warn "[$$] Disabling keep-alive after sending unknown length body on $protocol\n";
            $self->{client}->{keepalive} = 0;
        }
    }

    if ( ! $headers{date} ) {
        push @headers, "Date: " . time2str( time() );
    }

    # Should we keep the connection open?
    if ( $self->{client}->{keepalive} ) {
        push @headers, 'Connection: keep-alive';
    } else {
        push @headers, 'Connection: close';
    }

    my $conn = $self->{server}->{client};

    # Buffer the headers so they are sent with the first write() call
    # This reduces the number of TCP packets we are sending
    _syswrite($conn, \(join( $CRLF, @headers, '' ) . $CRLF));

    if (defined $res->[2]) {
        Plack::Util::foreach($res->[2], sub {
            my $buffer = $_[0];
            if ($chunked) {
                my $len = length $buffer;
                return unless $len;
                $buffer = sprintf( "%x", $len ) . $CRLF . $buffer . $CRLF;
            }
            _syswrite($conn, \$buffer);
        });
        _syswrite($conn, \"0$CRLF$CRLF") if $chunked;
    } else {
        return Plack::Util::inline_object
            write => sub {
                my $buffer = $_[0];
                if ($chunked) {
                    my $len = length $buffer;
                    return unless $len;
                    $buffer = sprintf( "%x", $len ) . $CRLF . $buffer . $CRLF;
                }
                _syswrite($conn, \$buffer);
            },
            close => sub {
                _syswrite($conn, \"0$CRLF$CRLF") if $chunked;
            };
    }
}

sub _syswrite {
    my ($conn, $buffer_ref) = @_;

    my $amount = length $$buffer_ref;
    my $offset = 0;

    while ($amount > 0) {
        my $len = syswrite($conn, $$buffer_ref, $amount, $offset);

        if (not defined $len) {
            return if $! == EPIPE;
            return if $! == ECONNRESET;
            redo if $! == EINTR;
            die "write error: $!";
        }

        $amount -= $len;
        $offset += $len;

        DEBUG && warn "[$$] Wrote $len byte", ($len == 1 ? '' : 's'), "\n";
    }
}

sub post_client_connection_hook {
    my $self = shift;
    if ($self->{client}->{harakiri}) {
        exit;
    }
}

1;
